staging/src/k8s.io/api/discovery/v1alpha1/types.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// EndpointSlice represents a subset of the endpoints that implement a service.
+// For a given service there may be multiple EndpointSlice objects, selected by
+// labels, which must be joined to produce the full set of endpoints.
+type EndpointSlice struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// endpoints is a list of unique endpoints in this slice. Each slice may
+	// include a maximum of 1000 endpoints.
+	// +listType=atomic
+	Endpoints []Endpoint `json:"endpoints" protobuf:"bytes,2,rep,name=endpoints"`
+	// ports specifies the list of network ports exposed by each endpoint in
+	// this slice. Each port must have a unique name. When ports is empty, it
+	// indicates that there are no defined ports. The most common example of
```

Comment:
Is "The most common example" needed here?  It seems to be emphasizing a relatively corner-case and exposing impl details about one potential consumer of this API.  We should instead just say what we expect consumers to do if the find >0 endpoints and 0 ports.  E.g. "..no defined ports.  Consumers should interpret this as 'all ports'."

This fails-open, but again with POLS I think this is probably right.  If there's a security concern with publishing ports (security through obscurity) you have a different problem.  How does that fit?

Reply:
> what we expect consumers to do if the find >0 endpoints and 0 ports. E.g. "..no defined ports. Consumers should interpret this as 'all ports'."

Can we do this? 

- **">0 enpdoints and 0 ports"**: Up to the consumer to interpret. (Ports may be irrelevant)
- **">0 endpionts and 1 EndpointPort with nil port"**: Consumer must interpret as "all ports"

Reply:
That seems fine.  We can smooth out exact wording once it is in.

Specifically, no ports[] means both protocols and ports are not defined - it's an address with no other information.  ports[] with no port number means "all ports for the specified protocol", because protocol will be defaulted.

The point here is not to leave the meaning of these cases up for debate.